File: library/src/main/java/com/imangazaliev/circlemenu/CircleMenu.java
package com.imangazaliev.circlemenu;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.List;

public class CircleMenu extends FrameLayout implements MenuControllerListener {

    private int distance;
    private int circleStartAngle;
    private int angleRange;
    private boolean openOnStart;
    private boolean hintsEnabled;
    private int calculatedSize;
    private int centerButtonGravity;

    private CenterMenuButton centerButton;
    private MenuController menuController;

    private EventListener eventListener;
    private OnItemClickListener onItemClickListener;

    public CircleMenu(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public CircleMenu(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        init(attrs);
    }

    private void init(AttributeSet attrs) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.CircleMenu);

        try {
            circleStartAngle = typedArray.getInteger(R.styleable.CircleMenu_startAngle, getResources().getInteger(R.integer.circle_menu_start_angle));
            angleRange = typedArray.getInteger(R.styleable.CircleMenu_angleRange, 360);
            distance = (int) typedArray.getDimension(R.styleable.CircleMenu_distance, getResources().getDimension(R.dimen.circle_menu_distance));
            centerButtonGravity = typedArray.getInt(R.styleable.CircleMenu_centerButtonGravity, Gravity.CENTER);
            openOnStart = typedArray.getBoolean(R.styleable.CircleMenu_openOnStart, false);
            hintsEnabled = typedArray.getBoolean(R.styleable.CircleMenu_hintsEnabled, false);
        } finally {
            typedArray.recycle();
        }

        float buttonSize = getResources().getDimension(R.dimen.circle_menu_button_size);
        int ringRadius = (int) (buttonSize + (distance - buttonSize / 2));
        calculatedSize = (int) (ringRadius * 2 * ItemSelectionAnimator.END_CIRCLE_SIZE_RATIO);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        createCenterButton();
    }

    private void createCenterButton() {
        centerButton = new CenterMenuButton(getContext(), openOnStart);
        centerButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                toggle();
            }
        });
        LayoutParams centerButtonParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        centerButtonParams.gravity = centerButtonGravity;

        centerButton.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                centerButton.getViewTreeObserver().removeOnGlobalLayoutListener(this);

                initMenuController();
            }
        });

        addView(centerButton, centerButtonParams);
    }

    private void initMenuController() {
        List<CircleMenuButton> buttons = new ArrayList<>();
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            if (child != centerButton) {
                buttons.add((CircleMenuButton) child);
            }
        }

        menuController = new MenuController(getContext(), buttons, this, centerButton.getX(), centerButton.getY(),
                circleStartAngle, angleRange, distance, openOnStart, hintsEnabled);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        final int w = resolveSizeAndState(calculatedSize, widthMeasureSpec, 0);
        final int h = resolveSizeAndState(calculatedSize, heightMeasureSpec, 0);

        setMeasuredDimension(w, h);
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        super.dispatchDraw(canvas);

        menuController.onDraw(canvas);
    }

    @Override
    public void onOpenAnimationStart() {
        centerButton.setClickable(false);
        centerButton.setOpened(true);

        if (eventListener != null) {
            eventListener.onMenuOpenAnimationStart();
        }
    }

    @Override
    public void onOpenAnimationEnd() {
        centerButton.setClickable(true);

        if (eventListener != null) {
            eventListener.onMenuOpenAnimationEnd();
        }
    }

    @Override
    public void onCloseAnimationStart() {
        centerButton.setClickable(false);
        centerButton.setOpened(false);

        if (eventListener != null) {
            eventListener.onMenuCloseAnimationStart();
        }
    }

    @Override
    public void onCloseAnimationEnd() {
        centerButton.setClickable(true);

        if (eventListener != null) {
            eventListener.onMenuCloseAnimationEnd();
        }
    }

    @Override
    public void onSelectAnimationStart(CircleMenuButton menuButton) {
        centerButton.setOpened(false);
        centerButton.setClickable(false);

        if (eventListener != null) {
            eventListener.onButtonClickAnimationStart(menuButton);
        }
    }

    @Override
    public void onSelectAnimationEnd(CircleMenuButton menuButton) {
        centerButton.setClickable(true);
        if (eventListener != null) {
            eventListener.onButtonClickAnimationEnd(menuButton);
        }
    }

    @Override
    public void redrawView() {
        invalidate();
    }

    @Override
    public void onClick(CircleMenuButton menuButton) {
        if (onItemClickListener != null) {
            onItemClickListener.onItemClick(menuButton);
        }
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void setEventListener(EventListener listener) {
        this.eventListener = listener;
    }

    public void toggle() {
        menuController.toggle();
    }

    public void open(boolean animate) {
        menuController.open(animate);
    }

    public void close(boolean animate) {
        menuController.close(animate);
    }

    public boolean isOpened() {
        return menuController.isOpened();
    }

    public interface OnItemClickListener {
        void onItemClick(CircleMenuButton menuButton);
    }

    public interface EventListener {

        void onMenuOpenAnimationStart();

        void onMenuOpenAnimationEnd();

        void onMenuCloseAnimationStart();

        void onMenuCloseAnimationEnd();

        void onButtonClickAnimationStart(@NonNull CircleMenuButton menuButton);

        void onButtonClickAnimationEnd(@NonNull CircleMenuButton menuButton);
    }


}

File: library/src/main/java/com/imangazaliev/circlemenu/CircleMenu.kt
package com.imangazaliev.circlemenu

import android.app.Activity
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Point
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton

class CircleMenu @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : FloatingActionButton(context, attrs) {

    val isOpened: Boolean
        get() = menuLayout.isOpened

    private val menuLayout: CircleMenuLayout

    init {
        isClickable = true

        context.obtainStyledAttributes(attrs, R.styleable.CircleMenu).apply {
            val circleStartAngle = getInteger(
                    R.styleable.CircleMenu_startAngle,
                    resources.getInteger(R.integer.circle_menu_start_angle)
            )
            val circleMaxAngle = getInteger(
                    R.styleable.CircleMenu_maxAngle,
                    resources.getInteger(R.integer.circle_menu_max_angle)
            )
            val distance = getDimension(
                    R.styleable.CircleMenu_distance,
                    resources.getDimension(R.dimen.circle_menu_distance)
            ).toInt()
            val openOnStart = getBoolean(R.styleable.CircleMenu_openOnStart, false)

            val centerButtonColorDef = ContextCompat.getColor(context, R.color.circle_menu_center_button_color)
            val centerButtonColor = getColor(R.styleable.CircleMenu_centerButtonColor, centerButtonColorDef)
            val centerButtonIconColorDef = ContextCompat.getColor(context, R.color.circle_menu_center_button_icon_color)
            val centerButtonIconColor = getColor(R.styleable.CircleMenu_centerButtonIconColor, centerButtonIconColorDef)
            val menuIconType = MenuIconType.values()[getInt(R.styleable.CircleMenu_menuIcon, 0)]

            val iconsColorDef = ContextCompat.getColor(context, R.color.circle_menu_button_icon_color)
            val buttonIconsColor = getColor(R.styleable.CircleMenu_iconsColor, iconsColorDef)
            val iconArrayId: Int = getResourceId(R.styleable.CircleMenu_buttonIcons, 0)
            val colorArrayId: Int = getResourceId(R.styleable.CircleMenu_buttonColors, 0)

            val showSelectAnimation: Boolean = getBoolean(R.styleable.CircleMenu_showSelectAnimation, true)

            val colors = resources.getIntArray(colorArrayId).asList()
            val icons = resources.obtainTypedArray(iconArrayId).let { iconsIds ->
                (0 until iconsIds.length()).map {
                    iconsIds.getResourceId(it, -1)
                }
            }

            if (colors.isEmpty() || icons.isEmpty()) {
                throw IllegalArgumentException("Colors and icons array must not be empty")
            }

            if (colors.size != icons.size) {
                throw IllegalArgumentException("Colors array size must be equal to the icons array")
            }

            menuLayout = CircleMenuLayout(
                    context = context,
                    centerButtonColor = centerButtonColor,
                    centerButtonIconColor = centerButtonIconColor,
                    menuIconType = menuIconType,
                    buttonIconsColor = buttonIconsColor,
                    distance = distance,
                    circleMaxAngle = circleMaxAngle,
                    circleStartAngle = circleStartAngle,
                    showSelectAnimation = showSelectAnimation,
                    openOnStart = openOnStart,
                    colors = colors,
                    icons = icons
            )

            initMenuButton(menuIconType, centerButtonColor, centerButtonIconColor)
            initMenuLayout()
        }.recycle()
    }

    private fun initMenuButton(
            iconType: MenuIconType,
            buttonColor: Int,
            iconColor: Int
    ) {
        val iconResId = if (iconType == MenuIconType.PLUS) {
            R.drawable.ic_plus
        } else {
            R.drawable.ic_menu
        }
        val icon = ContextCompat.getDrawable(context, iconResId)!!
        icon.setTintCompat(iconColor)
        setImageDrawable(icon)
        backgroundTintList = ColorStateList.valueOf(buttonColor)
    }

    private fun initMenuLayout() {
        val activity = context as Activity
        val decor = activity.window.decorView as ViewGroup
        val menuButton = this

        onLaidOut {
            val menuButtonLocation = IntArray(2).let {
                menuButton.getLocationOnScreen(it)
                Point(it.first(), it.last())
            }

            val bounds = Rect(menuButtonLocation.x, menuButtonLocation.y,
                    menuButtonLocation.x + menuButton.width,
                    menuButtonLocation.y + menuButton.height
            )

            val parentLocation = IntArray(2).let {
                decor.getLocationOnScreen(it)
                Point(it.first(), it.last())
            }
            bounds.offset(-parentLocation.x, -parentLocation.y)

            val menuLayoutParams = FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.WRAP_CONTENT,
                    FrameLayout.LayoutParams.WRAP_CONTENT
            )
            decor.addView(menuLayout, menuLayoutParams)
            menuLayout.post {
                val width = menuLayout.width
                val height = menuLayout.height
                decor.removeView(menuLayout)
                val params = FrameLayout.LayoutParams(
                        width,
                        height
                )
                params.leftMargin = bounds.centerX() - (menuLayout.width / 2)
                params.topMargin = bounds.centerY() - (menuLayout.height / 2)
                decor.addView(menuLayout, params)
            }
        }



        super.setOnClickListener {
            showMenu()
        }
    }

    private fun showMenu() {
        menuLayout.visibility = View.VISIBLE
        menuLayout.open(true)
    }

    override fun setOnClickListener(l: OnClickListener?) {
        //empty
    }

    fun setOnItemClickListener(listener: (buttonIndex: Int) -> Unit) {
        this.menuLayout.setOnItemClickListener(listener)
    }

    fun setOnItemLongClickListener(listener: (buttonIndex: Int) -> Unit) {
        this.menuLayout.setOnItemLongClickListener(listener)
    }

    fun toggle() {
        this.menuLayout.toggle()
    }

    fun open(animate: Boolean = true) {
        this.menuLayout.open(animate)
    }

    fun close(animate: Boolean) {
        this.menuLayout.open(animate)
    }

    fun onMenuOpenAnimationStart(listener: () -> Unit) {
        this.menuLayout.onMenuOpenAnimationStart(listener)
    }

    fun onMenuOpenAnimationEnd(listener: () -> Unit) {
        this.menuLayout.onMenuOpenAnimationEnd(listener)
    }

    fun onMenuCloseAnimationStart(listener: () -> Unit) {
        this.menuLayout.onMenuCloseAnimationStart(listener)
    }

    fun onMenuCloseAnimationEnd(listener: () -> Unit) {
        this.menuLayout.onMenuCloseAnimationEnd(listener)
    }

    fun onButtonClickAnimationStart(listener: (buttonIndex: Int) -> Unit) {
        this.menuLayout.onButtonClickAnimationStart(listener)
    }

    fun onButtonClickAnimationEnd(listener: (buttonIndex: Int) -> Unit) {
        this.menuLayout.onButtonClickAnimationEnd(listener)
    }

    internal enum class MenuIconType(val code: String) {
        HAMBURGER("hamburger"), PLUS("plus")
    }

}